dynix.3.2.0/src/doc/man1/dd.1
.\" $Copyright:	$
.\" Copyright (c) 1984, 1985, 1986, 1987, 1988, 1989, 1990 
.\" Sequent Computer Systems, Inc.   All rights reserved.
.\"  
.\" This software is furnished under a license and may be used
.\" only in accordance with the terms of that license and with the
.\" inclusion of the above copyright notice.   This software may not
.\" be provided or otherwise made available to, or used by, any
.\" other person.  No title to or ownership of the software is
.\" hereby transferred.
...
.V= $Header: dd.1 1.9 90/03/22 $
.TH DD 1 "\*(V)" "4BSD"
.SH NAME
\f1dd \- convert and copy a file\fP
.SH SYNOPSIS
\f3dd\f1
[\f3option=\f2value\f1] . . .
.SH DESCRIPTION
.B Dd
copies the specified input file
to the specified output with
possible conversions.
The standard input and output are used by default.
The input and output block size may be
specified to take advantage of raw physical I/O.\fP
.SH OPTIONS
.IP \f3if=\f1 15
Input filename; standard input is default
.IP \f3of=\f1 15
Output filename; standard output is default
.IP \f3ibs=\f1 15
\f2n\f1 input block size \f2n\f1 bytes (default 512)
.IP \f3obs=\f1 15
\f2n\f1 output block size (default 512)
.IP \f3bs=\fP 15
Set both input and output block size,
superseding \f3ibs\f1 and \f3obs\f1.  If no conversion is specified,
this function is particularly efficient, because there is no need to
copy.
.IP \f3cbs=\fP 15
Conversion buffer size
.IP \f3skip=\fP 15
Skip \f2n\f1 input records before starting copy
.IP \f3files=\fP 15
Copy \f2n\f1 input files before terminating (where input is a magtape
or similar device)
.IP \f3seek=\fP 15
Seek \f2n\f1 records from beginning of output file before copying
.IP \f3count=\fP 15
Copy only \f2n\f1 input records
.PP
The following conversion options are specified by \f3conv=\f1[\f2option\f1]:
.IP \f3ascii\f1 15
Convert EBCDIC to ASCII
.IP \f3ebcdic\f1 
Convert ASCII to EBCDIC
.IP \f3ibm\f1
Slightly different map of ASCII to EBCDIC
.IP \f3block\f1
Convert variable length records to fixed length
.IP \f3unblock\f1
Convert fixed length records to variable length
.IP \f3lcase\f1 
Map alphabetics to lower case
.IP \f3ucase\f1
Map alphabetics to upper case
.IP \f3swab\f1
Swap every pair of bytes
.IP \f3noerror\f1
Do not stop processing on an error
.IP \f3sync\f1
Pad every input record to \f3ibs\f1
.IP "\f3 . . . , . . . "\f1
Several comma-separated conversions
.PP
Where sizes are specified,
a number of bytes is expected.
A number may end with
.B "k, b"
or
.B w
to specify multiplication by
1024, 512, or sizeof(\f2int\fP);
a pair of numbers may be separated by
.B x
to indicate a product.
.PP
.B Cbs
is used only if
.B ascii,
.B unblock,
.B ebcdic,
.B ibm,
or
.B block
conversion is specified.
In the first two cases,
.B cbs
characters are placed into the conversion buffer, any specified
character mapping is done,
trailing blanks are trimmed, and newline is added
before sending the line to the output.
In the latter three cases, characters are read into the
conversion buffer and blanks are added
to make up an
output record of size
.BR cbs .
.PP
After completion,
.B dd
reports the number of whole and partial input and output
blocks.
.SH EXAMPLES
To read an EBCDIC tape, blocked ten 80-byte
EBCDIC card images per record, into ASCII file
.I x,
type this command:\fP
.IP
.nf
\f3dd if\f1=\f2/dev/rmt0\fP \f3of\f1=\f2x\f1 \f3ibs=\f2\f2800\f1
\f3cbs=\f\f2180\f1 \f3conv=\f1\f3ascii,lcase\f1
.fi
.PP
Note that raw magtape is used.
.B Dd
is especially suited to I/O on raw
physical devices, because it allows reading
and writing in arbitrary record sizes.
.SH "SEE ALSO"
\f1cp(1), tr(1)\fP
.SH DIAGNOSTICS
.ft CW
f+p records in(out): numbers of full and partial records read(written)\fP
.SH BUGS
The ASCII/EBCDIC conversion tables are
taken
from the 256-character standard in
the \f2CACM\f1, Nov, 1968.
The ibm conversion, while less accepted as a standard,
corresponds better to certain IBM print train conventions.
There is no universal solution.
.PP
To ensure that \f3dd\f1 stays synchronized, specify the following
when copying raw disks with bad sectors:
.IP
.B conv=noerror,sync


dynix.3.2.0/src/5cmd/dd.c
/*	@(#)dd.c	1.8	*/
/*
**	convert and copy
*/

#include	<stdio.h>
#include	<sys/param.h>
#include	<signal.h>
#include	<sys/types.h>
#include	<sys/sysmacros.h>

#define	LCASE	01
#define	UCASE	02
#define	SWAB	04
#define NERR	010
#define SYNC	020

/* If aligned for 4K pages, also aligned for 512 */
#define PAGESIZE 4096

unsigned ibs = BSIZE;
unsigned obs = BSIZE;
unsigned bs;
unsigned cbs;
unsigned ibc;
unsigned obc;
unsigned cbc;

int	fflag;
int	cflag;
int	nifr;
int	nipr;
int	nofr;
int	nopr;
int	ntrunc;
int	ibf;
int	obf;
int	skip;
int	seekn;
int	count;
int	nspace;

char	*string;
char	*ifile;
char	*ofile;
char	*ibuf;
char	*obuf;
char	*sbrk();
char	*op;

/* This is an EBCDIC to ASCII conversion table
*/
char etoa [] =
{
	0000,0001,0002,0003,0234,0011,0206,0177,
	0227,0215,0216,0013,0014,0015,0016,0017,
	0020,0021,0022,0023,0235,0205,0010,0207,
	0030,0031,0222,0217,0034,0035,0036,0037,
	0200,0201,0202,0203,0204,0012,0027,0033,
	0210,0211,0212,0213,0214,0005,0006,0007,
	0220,0221,0026,0223,0224,0225,0226,0004,
	0230,0231,0232,0233,0024,0025,0236,0032,
	0040,0240,0241,0242,0243,0244,0245,0246,
	0247,0250,0325,0056,0074,0050,0053,0174,
	0046,0251,0252,0253,0254,0255,0256,0257,
	0260,0261,0041,0044,0052,0051,0073,0176,
	0055,0057,0262,0263,0264,0265,0266,0267,
	0270,0271,0313,0054,0045,0137,0076,0077,
	0272,0273,0274,0275,0276,0277,0300,0301,
	0302,0140,0072,0043,0100,0047,0075,0042,
	0303,0141,0142,0143,0144,0145,0146,0147,
	0150,0151,0304,0305,0306,0307,0310,0311,
	0312,0152,0153,0154,0155,0156,0157,0160,
	0161,0162,0136,0314,0315,0316,0317,0320,
	0321,0345,0163,0164,0165,0166,0167,0170,
	0171,0172,0322,0323,0324,0133,0326,0327,
	0330,0331,0332,0333,0334,0335,0336,0337,
	0340,0341,0342,0343,0344,0135,0346,0347,
	0173,0101,0102,0103,0104,0105,0106,0107,
	0110,0111,0350,0351,0352,0353,0354,0355,
	0175,0112,0113,0114,0115,0116,0117,0120,
	0121,0122,0356,0357,0360,0361,0362,0363,
	0134,0237,0123,0124,0125,0126,0127,0130,
	0131,0132,0364,0365,0366,0367,0370,0371,
	0060,0061,0062,0063,0064,0065,0066,0067,
	0070,0071,0372,0373,0374,0375,0376,0377,
};
/* This is an ASCII to EBCDIC conversion table
*/
char atoe [] =
{
0000,0001,0002,0003,0067,0055,0056,0057,
0026,0005,0045,0013,0014,0015,0016,0017,
0020,0021,0022,0023,0074,0075,0062,0046,
0030,0031,0077,0047,0034,0035,0036,0037,
0100,0132,0177,0173,0133,0154,0120,0175,
0115,0135,0134,0116,0153,0140,0113,0141,
0360,0361,0362,0363,0364,0365,0366,0367,
0370,0371,0172,0136,0114,0176,0156,0157,
0174,0301,0302,0303,0304,0305,0306,0307,
0310,0311,0321,0322,0323,0324,0325,0326,
0327,0330,0331,0342,0343,0344,0345,0346,
0347,0350,0351,0255,0340,0275,0232,0155,
0171,0201,0202,0203,0204,0205,0206,0207,
0210,0211,0221,0222,0223,0224,0225,0226,
0227,0230,0231,0242,0243,0244,0245,0246,
0247,0250,0251,0300,0117,0320,0137,0007,
0040,0041,0042,0043,0044,0025,0006,0027,
0050,0051,0052,0053,0054,0011,0012,0033,
0060,0061,0032,0063,0064,0065,0066,0013,
0070,0071,0072,0073,0004,0024,0076,0341,
0101,0102,0103,0104,0105,0106,0107,0110,
0111,0121,0122,0123,0124,0125,0126,0127,
0130,0131,0142,0143,0144,0145,0146,0147,
0150,0151,0160,0161,0162,0163,0164,0165,
0166,0167,0170,0200,0212,0213,0214,0215,
0216,0217,0220,0152,0233,0234,0235,0236,
0237,0240,0252,0253,0254,0112,0256,0257,
0260,0261,0262,0263,0264,0265,0266,0267,
0270,0271,0272,0273,0274,0241,0276,0277,
0312,0313,0314,0315,0316,0317,0332,0333,
0334,0335,0336,0337,0352,0353,0354,0355,
0356,0357,0372,0373,0374,0375,0376,0377,
};
char	atoibm[] =
{
	0000,0001,0002,0003,0067,0055,0056,0057,
	0026,0005,0045,0013,0014,0015,0016,0017,
	0020,0021,0022,0023,0074,0075,0062,0046,
	0030,0031,0077,0047,0034,0035,0036,0037,
	0100,0132,0177,0173,0133,0154,0120,0175,
	0115,0135,0134,0116,0153,0140,0113,0141,
	0360,0361,0362,0363,0364,0365,0366,0367,
	0370,0371,0172,0136,0114,0176,0156,0157,
	0174,0301,0302,0303,0304,0305,0306,0307,
	0310,0311,0321,0322,0323,0324,0325,0326,
	0327,0330,0331,0342,0343,0344,0345,0346,
	0347,0350,0351,0255,0340,0275,0137,0155,
	0171,0201,0202,0203,0204,0205,0206,0207,
	0210,0211,0221,0222,0223,0224,0225,0226,
	0227,0230,0231,0242,0243,0244,0245,0246,
	0247,0250,0251,0300,0117,0320,0241,0007,
	0040,0041,0042,0043,0044,0025,0006,0027,
	0050,0051,0052,0053,0054,0011,0012,0033,
	0060,0061,0032,0063,0064,0065,0066,0010,
	0070,0071,0072,0073,0004,0024,0076,0341,
	0101,0102,0103,0104,0105,0106,0107,0110,
	0111,0121,0122,0123,0124,0125,0126,0127,
	0130,0131,0142,0143,0144,0145,0146,0147,
	0150,0151,0160,0161,0162,0163,0164,0165,
	0166,0167,0170,0200,0212,0213,0214,0215,
	0216,0217,0220,0232,0233,0234,0235,0236,
	0237,0240,0252,0253,0254,0255,0256,0257,
	0260,0261,0262,0263,0264,0265,0266,0267,
	0270,0271,0272,0273,0274,0275,0276,0277,
	0312,0313,0314,0315,0316,0317,0332,0333,
	0334,0335,0336,0337,0352,0353,0354,0355,
	0356,0357,0372,0373,0374,0375,0376,0377,
};


int (*signal())();
main(argc, argv)
char **argv;
{
	register char *ip;
	register c;
	int	(*conv)();
	int	ebcdic(), ibm(), ascii(), null(), cnull(), term();
	int	a;

	conv = null;
	for(c=1; c<argc; c++) {
		string = argv[c];
		if(match("ibs=")) {
			ibs = number();
			continue;
		}
		if(match("obs=")) {
			obs = number();
			continue;
		}
		if(match("cbs=")) {
			cbs = number();
			continue;
		}
		if (match("bs=")) {
			bs = number();
			continue;
		}
		if(match("if=")) {
			ifile = string;
			continue;
		}
		if(match("of=")) {
			ofile = string;
			continue;
		}
		if(match("skip=")) {
			skip = number();
			continue;
		}
		if(match("seek=")) {
			seekn = number();
			continue;
		}
		if(match("count=")) {
			count = number();
			continue;
		}
		if(match("conv=")) {
		cloop:
			if(match(","))
				goto cloop;
			if(*string == '\0')
				continue;
			if(match("ebcdic")) {
				conv = ebcdic;
				goto cloop;
			}
			if(match("ibm")) {
				conv = ibm;
				goto cloop;
			}
			if(match("ascii")) {
				conv = ascii;
				goto cloop;
			}
			if(match("lcase")) {
				cflag |= LCASE;
				goto cloop;
			}
			if(match("ucase")) {
				cflag |= UCASE;
				goto cloop;
			}
			if(match("swab")) {
				cflag |= SWAB;
				goto cloop;
			}
			if(match("noerror")) {
				cflag |= NERR;
				goto cloop;
			}
			if(match("sync")) {
				cflag |= SYNC;
				goto cloop;
			}
		}
		fprintf(stderr,"dd: bad arg %s\n", string);
		exit(2);
	}
	if(conv == null && cflag&(LCASE|UCASE))
		conv = cnull;
	if (ifile)
		ibf = open(ifile, 0);
#ifndef STANDALONE
	else
		ibf = dup(0);
#endif
	if(ibf < 0) {
		fprintf(stderr,"dd: cannot open %s\n", ifile);
		exit(2);
	}
	if (ofile)
		obf = creat(ofile, 0666);
#ifndef STANDALONE
	else
		obf = dup(1);
#endif
	if(obf < 0) {
		fprintf(stderr,"dd: cannot create %s\n", ofile);
		exit(2);
	}
	if (bs) {
		ibs = obs = bs;
		if (conv == null)
			fflag++;
	}
	if(ibs == 0 || obs == 0) {
		fprintf(stderr,"dd: counts cannot be zero\n");
		exit(2);
	}
	ibuf = sbrk(ibs + PAGESIZE);
	/* page align */
	if (ibuf != (char *) -1) {
		ibuf = (char *) (((int)ibuf + (PAGESIZE - 1)) &
				 ~(PAGESIZE - 1));
	}
	if (fflag)
		obuf = ibuf;
	else {
		obuf = sbrk(obs + PAGESIZE);
		if (obuf != (char *) -1) {
			obuf = (char *) (((int)obuf + (PAGESIZE - 1)) &
					 ~(PAGESIZE - 1));
		}
	}
	sbrk((unsigned)64);	/* For good measure */
	if(ibuf == (char *)-1 || obuf == (char *)-1) {
		fprintf(stderr,"dd: not enough memory\n");
		exit(2);
	}
	ibc = 0;
	obc = 0;
	cbc = 0;
	op = obuf;

#ifndef STANDALONE
	if(signal(SIGINT, SIG_IGN) != SIG_IGN)
		signal(SIGINT, term);
#endif
	while(skip) {
		read(ibf, ibuf, ibs);
		skip--;
	}
	while(seekn) {
		lseek(obf, (long)obs, 1);
		seekn--;
	}

	while(1) {
		if(ibc-- == 0) {
			ibc = 0;
			if(count == 0 || nifr+nipr != count) {
				if(cflag&(NERR|SYNC))
				for(ip=ibuf+ibs; ip>ibuf;)
					*--ip = 0;
				ibc = read(ibf, ibuf, ibs);
			}
			if(ibc == -1) {
				perror("dd read error");
				if((cflag&NERR) == 0) {
					flsh();
					term(2);
				}
				ibc = 0;
				for(c=0; c<ibs; c++)
					if(ibuf[c] != 0)
						ibc = c;
				stats();
			}
			if(ibc == 0) {
				flsh();
				term(0);
			}
			if(ibc != ibs) {
				nipr++;
				if(cflag&SYNC)
					ibc = ibs;
			} else
				nifr++;
			ip = ibuf;
		/*	if (ibc % 2) ibuf[ibc] = 0;	*/
			c = (ibc + 1) >> 1;
			if(cflag&SWAB && c)
				do {
					a = *ip++;
					ip[-1] = *ip;
					*ip++ = a;
				} while(--c);
			ip = ibuf;
			if (fflag) {
				obc = ibc;
				flsh();
				ibc = 0;
			}
			continue;
		}
		c = 0;
		c |= *ip++;
		(*conv)(c);
	}
}

flsh()
{
	register c;

	if(obc) {
		if(obc == obs)
			nofr++;
		else
			nopr++;
		c = write(obf, obuf, obc);
		if(c != obc) {
			perror("dd write error");
			term(2);
		}
		obc = 0;
	}
}

match(s)
char *s;
{
	register char *cs;

	cs = string;
	while(*cs++ == *s)
		if(*s++ == '\0')
			goto true;
	if(*s != '\0')
		return(0);

true:
	cs--;
	string = cs;
	return(1);
}

number()
{
	register char *cs;
	register n;

	cs = string;
	n = 0;
	while(*cs >= '0' && *cs <= '9')
		n = n*10 + *cs++ - '0';
	for(;;)
	switch(*cs++) {

	case 'k':
		n *= 1024;
		continue;

	case 'w':
		n *= 2;
		continue;

	case 'b':
		n *= BSIZE;
		continue;

	case '*':
	case 'x':
		string = cs;
		n *= number();

	case '\0':
		return(n);
	}
	/* never gets here */
}

cnull(cc)
{
	register c;

	c = cc;
	if(cflag&UCASE && c>='a' && c<='z')
		c += 'A'-'a';
	if(cflag&LCASE && c>='A' && c<='Z')
		c += 'a'-'A';
	null(c);
}

null(c)
{

	*op = c;
	op++;
	if(++obc >= obs) {
		flsh();
		op = obuf;
	}
}

ascii(cc)
{
	register c;

	c = 0;
	c |= etoa[0377 & cc];
	if(cbs == 0) {
		cnull(c);
		return;
	}
	if(c == ' ')
		nspace++;
	else {
		while(nspace > 0) {
			null(' ');
			nspace--;
		}
		cnull(c);
	}
	if(++cbc >= cbs) {
		null('\n');
		cbc = 0;
		nspace = 0;
	}
}

ebcdic(cc)
{
	register c;

	c = cc;
	if(cflag&UCASE && c>='a' && c<='z')
		c += 'A'-'a';
	if(cflag&LCASE && c>='A' && c<='Z')
		c += 'a'-'A';
	c = atoe[c] & 0377;
	if(cbs == 0) {
		null(c);
		return;
	}
	if(cc == '\n') {
		while(cbc < cbs) {
			null(atoe[' ']);
			cbc++;
		}
		cbc = 0;
		return;
	}
	if(cbc == cbs)
		ntrunc++;
	cbc++;
	if(cbc <= cbs)
		null(c);
}

ibm(cc)
{
	register c;

	c = cc;
	if(cflag&UCASE && c>='a' && c<='z')
		c += 'A'-'a';
	if(cflag&LCASE && c>='A' && c<='Z')
		c += 'a'-'A';
	c = atoibm[c] & 0377;
	if(cbs == 0) {
		null(c);
		return;
	}
	if(cc == '\n') {
		while(cbc < cbs) {
			null(atoibm[' ']);
			cbc++;
		}
		cbc = 0;
		return;
	}
	if(cbc == cbs)
		ntrunc++;
	cbc++;
	if(cbc <= cbs)
		null(c);
}

term(c)
{

	stats();
	exit(c);
}

stats()
{

	fprintf(stderr,"%u+%u blocks in\n", nifr, nipr);
	fprintf(stderr,"%u+%u blocks out\n", nofr, nopr);
	if(ntrunc)
		fprintf(stderr,"%u truncated blocks\n", ntrunc);
}
